### DISCUSSION TITLE
Zugriff auf Dialog(ue)Trees bzw. dessen Resourcen

### DISCUSSION THREAD

**Discussion by Ize-dev** [Drova-Modding/Drova-Modding-API#12] [General]
Komme ich hiermit irgendwie an die DialogTrees für das gesamte Spiel?

Ich kann die zwar loggen aber das ist übelst Aufwand xD

Bis jetzt hab ich mir mit dem locPath und context.ToString() ausgeholfen das passt meistens.

Und einfach nur die .loc-Dateien parsen geht zwar aber teilweise "passen" die NPCs nicht zu dem NPC.loc unter dem der Text abgelegt ist.

Insbesondere im Intro springt das wild hin und her, da ist z.b. der Dialog von Venia aka Idefix in den Dateien von Maeron aka Miraculix (lol).
Bei den anderen NPCs ist z.B. auch immer der auswählbare Spielerdialog in den Dateien der NPCs gespeichert.

Da hilft es auch nicht, dass die Strings alle ähnliche Bezeichner mit Plh_ haben ;)

**Comment by TrustNoOneElse** [comment]
Hi @Ize-dev, es könnte sein das die Dialoge alle in den Addressables hinterlegt sind, schaue ich später mal nach, dann sollten sie leicht rauszufiltern sein :)
Was genau ist den dein Plan? Willst du selber welche erzeugen oder bearbeiten? Dann könnten wir schauen, dass wir dir da den Zugriff erleichtern^^
Schätze ich bin in ca. 30 Minuten vorm Laptop und kann dann auch detailliert antworten :D

**Reply by TrustNoOneElse** [reply]
Okay die Addressables beinhalten sehr viele `NodeCanvas.DialogueTrees.DialogueTree` mit dem jeweiligen NPC_Namen nach denen suchst du wahrscheinlich oder? 
https://github.com/Drova-Modding/Drova-Modding-API/wiki/Create-Addressable-Info-File-EN#wiki-wrapper
Du kannst dir auch jederzeit selbst so eine Addressable Datei generieren, wenn du magst, nur als Info ^^

**Comment by Ize-dev** [comment]
Danke, ich hab die Addressables mal generiert aber ich glaube das hilft mir nicht viel weiter oder ich hab nen Denkfehler (was gut möglich ist) :D

nodeCanvas ist was proprietäres https://nodecanvas.paradoxnotion.com was ich nicht einfach so auseinander nehmen kann.

Hatte gehofft da ist ein einfacher Ablauf drin den ich dann parsen kann.
Mir gehts rein um den Ablauf und wer was sagt.

Der Anfang vom Intro ist z.B.

| locDatei | Sprecher | Text |
|--------|--------|--------|
| IntroCutscene/Plh_0 | Player | Sie müssen irgendwo hier lang gelaufen sein. |
| IntroWorldDialog/Plh_0 | Miraculix | Wir müssen uns beeilen. |
| Dialog/NPC_Miraculix/Plh_12 | Miraculix | Die Welten berühren sich ... wir sind nah. |
| Dialog/NPC_Miraculix/Plh_19 | Idefix | Du hattest recht - es funktioniert! Das müssen wir den anderen berichten ... die werden Augen machen ... |
| Dialog/NPC_Miraculix/Plh_18 | Miraculix | Ob es wirklich funktioniert, werden wir erst wissen, wenn wir wirklich eine Verbindung zu Drova aufbauen ... |
| Dialog/NPC_Miraculix/Plh_41 | Idefix | Ist das nicht ... gefährlich? |

usw... das hab ich einfach gelogged.

Ist halt kreuz und quer durch die Dateien :)

**Reply by TrustNoOneElse** [reply]
Jap genau, aber das ist genau das was sie verwenden :D Eine Alternativ Idee wäre, dass du dir AssetRipper runterlädst und das ganze Projekt dir mal als Unity Projekt generieren lässt. Musst dir vorher allerdings noch die nightly MelonLoader Version holen um die Rewired Dll austauschen zu können, da ohne die Unity nicht starten wird ^^ Und wenn du Glück hast, solltest du darüber die Sachen dann leichter anschauen können ^^ Ich teste das ganze gerade im Background auch mal aus und informiere dich sobald es fertig ist :D

Sonst wird es denke ich halt leider nur den steinernen Weg und du musst das ganze halt leider doch selber parsen. Mit loggen wirst du ja leider nicht sehr weit kommen, beziehungsweise es wird sehr schwer, jeden case und jedes Gespräch damit aufzuzeichnen ^^

Als 3 Alternative gibt es auch den modding channel im Discord, wenn du Glück hast hat ja einer der Entwickler ein Tool, was sowas direkt ausspuckt :D

**Reply by TrustNoOneElse** [reply]
Also man kommt damit an die Grundstruktur dran, aber so wirklich was anfangen lässt sich damit erstmal nicht viel, ausser das man weiss welche Objekte referenziert sind. Werde morgen mal tiefer reinschauen, wie dieses nodecanvas funktioniert :D

**Reply by Ize-dev** [reply]
Danke für den Aufwand :D
Ich hab auch schon mit UnityExplorer, Hooks und der C# Konsole rumgespielt um evtl. damit was rauszuhauen aber hab nur mehrfach Drova erfolgreich gecrashed xD
Ich hab meine Zeit auf der Arbeit heute morgen aber schon sinnvoll genutzt und bin was auf der Spur:
https://nodecanvas.paradoxnotion.com/documentation/?section=json-importexport :D
Ich hoffe das ist kein Editorscript das nicht mit compiled wurde, dann sollte sich da was machen lassen... such ich heute Abend mal nach :)

**Reply by TrustNoOneElse** [reply]
Ja nicht dafür :D Das habe ich tatsächlich gestern auch indirekt gesehen, der `Graph` hat eine Funktion `getSerializedDataToJson` aber die gibt leider immer ein leeren Steing zurück.. hätte man deren Extension gekauft wäre es aber möglich einfach den Quellcode zu kopieren und dann es einfach wieder reinzupatchen, damit dann alles in json zu exportieren, was das ganze deutlich leichter machen würde :D

**Reply by Ize-dev** [reply]
Ich hab noch:

`string NodeCanvas.Framework.Graph.Serialize (bool pretyJson, List< UnityEngine.Object > objectReferences )`

in der Doku gefunden, kann aber erst heute Abend schauen ob das verfügbar ist und dann auch was ausgibt.

**Reply by TrustNoOneElse** [reply]
Die Methode ist leider gestripped mit diesen Params das was es noch gibt ist die ohne `bool prettyJson`, aber die gibt auch nur ein leeren String zurück.
Die einzigen Serialize Sache die funktioniert ist:
```cs
 var tree = Addressables.LoadAssetAsync<Il2CppNodeCanvas.DialogueTrees.DialogueTree>(assetReference).WaitForCompletion();
tree.GetSerializedReferencesBytesData();
tree
```
Beziehungsweise, es gibt auch noch andere Namen, aber sie führen alle zu dem gleichen Ergebnis wie von `GetSerializedReferencesBytesData` wo man die `List<UnityEngine.Object>` rausbekommt

**Reply by TrustNoOneElse** [reply]
Schade das dieses Asset einfach 90€ kostet... sonst hätte ich echt überlegt mir das zu holen, damit könnte man so viel leichter Sachen herausfinden :D

**Reply by TrustNoOneElse** [reply]
Alsooo an sich kommt man doch recht "einfach" ran, auf der "Scene_Actors" findet man denke ich alle Dialoge unter dem NPC jeweils. Das Problem scheint nur zu sein, dass diese Lazy loaded sind, deswegen hilft es auch nichts, die über die Addressables zu laden, da diese nicht initialisiert sind. Wenn man wüsste, wie genau diese geladen werden, könnte man die `_allNodes` durchgehen und mit dem `Il2CppNodeCanvas.DialogueTrees.DS_Statement` locakey und path, actorName, die Informationen rausbekommen :D Allerdings gibt es die auch als nested Variante, was das ganze noch ein wenig komplizierter macht, durch die MultipleChoice Nodes, soweit ich das gesehen habe.

**Comment by Ize-dev** [comment]
Eijeijei... ich schaus mir wahrscheinlich erst am Wochenende mal in Ruhe an... irgendwie wird sich da schon was machen lassen :D

**Reply by TrustNoOneElse** [reply]
Haha ja xD Also ich denke das einzig schwierige ist herauszufinden wie sie initalized werden, danach ist es nur noch ein bissl coding :D Im Notfall kann man sich ja sonst auch zu den Npc teleportieren und es so initialisieren, wenn man keinen einfacheren Weg findet^^

**Comment by TrustNoOneElse** [comment]
Hi @Ize-dev, ich bin jetzt an einem Punkt wo ich dabei bin selber Dialogue zu erstellen und einen Editor um welche anzulegen/bearbeiten und verstehe sie mittlerweile auch recht gut, falls du da noch dran bist und irgendwelche Infos brauchst, sag gerne bescheid :)
